### PR TITLE
Add 'indirect_sort'

### PR DESCRIPTION
**Pull request by mclow**

### CODE REVIEW

File: doc/indirect_sort.qbk [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+[/license
+Copyright (c) 2023 Marshall Clow
+
+Distributed under the Boost Software License, Version 1.0.
+(See accompanying file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+]
+
+There are times that you want a sorted version of a sequence, but for some reason or another, you don't really want to sort them.  Maybe the elements in the sequence are non-copyable (or non-movable), or the sequence is const, or they're just really expensive to move around.   An example of this might be a sequence of records from a database.
+
+Nevertheless, you might want to sort them.  That's where indirect sorting comes in.  In a "normal" sort, the elements of the sequence to be sorted are shuffled in place.  In indirect sorting, the elements are unchanged, but the sort algorithm returns to you a "permutation" of the elements that, when applied, will leave the elements in the sequence in a sorted order.
```
Comment (Flamefire) [inline doc/indirect_sort.qbk:14]: How about this a bit shorter wording especially avoiding to mention the need to sort twice:

```suggestion
There are times that you want a sorted version of a sequence, but for some reason you don't want to modify it.  Maybe the elements in the sequence can't be moved/copied, e.g. the sequence is const, or they're just really expensive to move around.   An example of this might be a sequence of records from a database.

That's where indirect sorting comes in.  In a "normal" sort, the elements of the sequence to be sorted are shuffled in place.  In indirect sorting, the elements are unchanged, but the sort algorithm returns a "permutation" of the elements that, when applied, will put the elements in the sequence in a sorted order.
```

Are the double-spaces after each sentence intended?

File: doc/indirect_sort.qbk [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+
+Say you have a sequence `[first, last)` of 1000 items that are expensive to swap:
```
Comment (Flamefire) [inline doc/indirect_sort.qbk:16]:
```suggestion
Assume a sequence `[first, last)` of 1000 items that are expensive to swap:
```

File: include/boost/algorithm/indirect_sort.hpp [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+/// \brief indirect sorting algorithms
+/// \author Marshall Clow
+///
+
+#ifndef BOOST_ALGORITHM_IS_INDIRECT_SORT
+#define BOOST_ALGORITHM_IS_INDIRECT_SORT
+
+#include <algorithm>        // for std::sort (and others)
+#include <functional>       // for std::less
+#include <vector>           // for std:;vector
```
Comment (Flamefire) [inline include/boost/algorithm/indirect_sort.hpp:19]: Typo: 
```suggestion
#include <vector>           // for std::vector
```

But is that comment really required?

File: include/boost/algorithm/indirect_sort.hpp [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+  
+*/
+
+/// \file indirect_sort.hpp
+/// \brief indirect sorting algorithms
+/// \author Marshall Clow
+///
+
+#ifndef BOOST_ALGORITHM_IS_INDIRECT_SORT
+#define BOOST_ALGORITHM_IS_INDIRECT_SORT
```
Comment (Flamefire) [inline include/boost/algorithm/indirect_sort.hpp:15]: Unusual include guard. Why not `BOOST_ALGORITHM_INDIRECT_SORT`?

File: include/boost/algorithm/indirect_sort.hpp [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+	boost::algorithm::iota(ret.begin(), ret.end(), size_t(0));
+	std::sort(ret.begin(), ret.end(), 
+	                  detail::indirect_predicate<Pred, RAIterator>(pred, first));
+	return ret;
+}
+
+/// \fn indirect_sort (RAIterator first, RAIterator las )
+/// \returns a permutation of the elements in the range [first, last)
+///     such that when the permutation is applied to the sequence,
+///     the result is sorted according to the predicate pred.
```
Comment (Flamefire) [inline include/boost/algorithm/indirect_sort.hpp:67]:
```suggestion
///     the result is sorted in non-descending order.
```

File: include/boost/algorithm/indirect_sort.hpp [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+template <typename RAIterator, typename Pred>
+std::vector<size_t> indirect_sort (RAIterator first, RAIterator last, Pred pred) {
+	Permutation ret(std::distance(first, last));
+	boost::algorithm::iota(ret.begin(), ret.end(), size_t(0));
+	std::sort(ret.begin(), ret.end(), 
+	                  detail::indirect_predicate<Pred, RAIterator>(pred, first));
+	return ret;
+}
+
+/// \fn indirect_sort (RAIterator first, RAIterator las )
```
Comment (Flamefire) [inline include/boost/algorithm/indirect_sort.hpp:64]:
```suggestion
/// \fn indirect_sort (RAIterator first, RAIterator last)
```

File: include/boost/algorithm/indirect_sort.hpp [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+/// \returns a permutation of the elements in the range [first, last)
+///     such that when the permutation is applied to the sequence,
+///     the result is sorted according to the predicate pred.
+///
+/// \param first  The start of the input sequence
+/// \param last   The end of the input sequence
+/// \param pred   The predicate to compare elements with
+///
+template <typename RAIterator, typename Pred>
+std::vector<size_t> indirect_sort (RAIterator first, RAIterator last, Pred pred) {
```
Comment (Flamefire) [inline include/boost/algorithm/indirect_sort.hpp:56]:
```suggestion
Permutation indirect_sort (RAIterator first, RAIterator last, Pred pred) {
```

File: include/boost/algorithm/indirect_sort.hpp [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+/// \fn indirect_sort (RAIterator first, RAIterator las )
+/// \returns a permutation of the elements in the range [first, last)
+///     such that when the permutation is applied to the sequence,
+///     the result is sorted according to the predicate pred.
+///
+/// \param first  The start of the input sequence
+/// \param last   The end of the input sequence
+///
+template <typename RAIterator>
+std::vector<size_t> indirect_sort (RAIterator first, RAIterator last) {
```
Comment (Flamefire) [inline include/boost/algorithm/indirect_sort.hpp:73]:
```suggestion
Permutation indirect_sort (RAIterator first, RAIterator last) {
```

File: test/indirect_sort_test.cpp [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+
+//	Make a copy of the data, apply the permutation, and ensure that it is sorted.
+	std::vector<typename std::iterator_traits<Iter>::value_type> v(first, last);
+	boost::algorithm::apply_permutation(v.begin(), v.end(), perm.begin(), perm.end());
+	BOOST_CHECK (boost::algorithm::is_sorted(v.begin(), v.end(), comp));
+}
+
+
+void test_sort () {
+    BOOST_CXX14_CONSTEXPR int num[] = { 1,3,5,7,9, 2, 4, 6, 8, 10 };
```
Comment (Flamefire) [inline test/indirect_sort_test.cpp:76]:
```suggestion
    int num[] = { 1,3,5,7,9, 2, 4, 6, 8, 10 };
```
or `int       *first  = &num[0];` is invalid isn't it?

File: test/indirect_sort_test.cpp [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+#include <iostream>
+#include <string>
+#include <vector>
+#include <list>
+
+typedef std::vector<size_t> Permutation;
+
+// A permutation of size N is a sequence of values in the range [0..N)
+// such that no value appears more than once in the permutation.
+bool isa_permutation(Permutation p, size_t N) {
```
Comment (Flamefire) [inline test/indirect_sort_test.cpp:27]:
```suggestion
bool is_a_permutation(Permutation p, size_t N) {
```
is more readable.

File: test/indirect_sort_test.cpp [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+    	test_one_sort(cFirst, cFirst + i, std::greater<int>());
+    	}
+    	
+//	make sure we work with iterators as well as pointers
+	std::vector<int> v(first, first + sz);
+	test_one_sort(v.begin(), v.end());
+	test_one_sort(v.begin(), v.end(), std::greater<int>());
+    }
+    
+BOOST_AUTO_TEST_CASE( test_main )
```
Comment (Flamefire) [inline test/indirect_sort_test.cpp:97]: Why that extra method and not using `BOOST_AUTO_TEST_CASE(test_sort)` directly?

File: test/indirect_sort_test.cpp [hunk 12]
```diff
@@ -0,0 +1,100 @@
+/* 
+   Copyright (c) Marshall Clow 2011-2012.
```
Comment (Flamefire) [inline test/indirect_sort_test.cpp:2]:
```suggestion
   Copyright (c) Marshall Clow 2023.
```

**Review by Flamefire — COMMENTED** [review]
Looks potentially useful.

Added some ideas for improving the description and pointed out a few places for improvement/typo-fixing.

File: test/indirect_sort_test.cpp (hunk 11 above)
Comment (mclow) [reply test/indirect_sort_test.cpp:97]: Because I expect there to be more test cases in the future.

Comment (Flamefire) [reply test/indirect_sort_test.cpp:97]: But the whole idea of `BOOST_AUTO_TEST_CASE` is that you simply "decorate" each test case with that and NOT have a "main" function. By default it will run each such function sequentially even allowing you to filter test based on their name from the CLI.

--> 

```
BOOST_AUTO_TEST_CASE( test_sort ){
...
}

BOOST_AUTO_TEST_CASE( test_indirect_stable_sort ){
...
}
```

File: include/boost/algorithm/indirect_sort.hpp [hunk 13]
```diff
@@ -0,0 +1,192 @@
...
+                                   RAIterator last, Pred pred) {
+	Permutation ret(std::distance(first, last));
+	
+	boost::algorithm::iota(ret.begin(), ret.end(), size_t(0));
+	std::partial_sort(ret.begin(), ret.begin() + std::distance(first, middle), ret.end(), 
+	                  detail::indirect_predicate<Pred, RAIterator>(pred, first));
+	return ret;
+}
+
+/// \fn indirect_partial_sort (RAIterator first, RAIterator last)
```
Comment (Flamefire) [inline include/boost/algorithm/indirect_sort.hpp:137]: C&P mistake in the signature inside this (and below) docstrings

File: include/boost/algorithm/indirect_sort.hpp [hunk 14]
```diff
@@ -0,0 +1,192 @@
...
+    return indirect_partial_sort(first, middle, last, 
+                    std::less<typename std::iterator_traits<RAIterator>::value_type>());
+}
+
+                    // ===== nth_element =====
+
+/// \fn indirect_nth_element (RAIterator first, RAIterator last, Predicate p)
+/// \returns a permutation of the elements in the range [first, last)
+///     such that when the permutation is applied to the sequence,
+///     the result is sorted according to the predicate pred.
```
Comment (Flamefire) [inline include/boost/algorithm/indirect_sort.hpp:158]: Similar C&P mistake in signature and description.

**Review by Flamefire — CHANGES_REQUESTED** [review]

File: include/boost/algorithm/indirect_sort.hpp (hunk 14 above)
Comment (mclow) [reply include/boost/algorithm/indirect_sort.hpp:158]: Nice catch - thanks!

**Comment by nigels-com** [comment]
I've implemented something similar for a recent project.
A few thoughts to note.

1.  Would like to used uint32_t for "small" jobs less than ~4 billion items
2. Would like to use this in combination with other sort algorithms such as boost::sort
3. Would like to use pre-existing permutation as an option, rather than always re-allocate one
